### ISSUE TITLE
Questions from WoN exploration

### ISSUE THREAD

**Issue by fosterlynn** [valueflows/valueflows#636]
From @fkleedorfer in the welcome chat.  [edit] Maybe we can just keep a running list of smaller questions here?

>For example: an event with move action has either one or two resources attached, If it has only one, it a) must not have a quantity, or b) may have no quantity, or its quantity must be equal to the resource's onhandQuantity, or c) it must have a quantity that is equal to the resource's onhand quantity?

>Or another: vf:primaryAccountable seems to be the relationship affected by events with vf:transfer and vf:transfer-all-rights actions.
Is there a relationship of the resource that is affected by vf:transfer-custody?

**Comment by fosterlynn** [comment]
>an event with move action has either one or two resources attached, If it has only one, it a) must not have a quantity, or b) may have no quantity, or its quantity must be equal to the resource's onhandQuantity, or c) it must have a quantity that is equal to the resource's onhand quantity?

Hmmm.  I think c) is best, it is good for events to always have a quantity.  And more consistent with moving part of a resource.  I assume this scenario means you just want to change the currentLocation.

>vf:primaryAccountable seems to be the relationship affected by events with vf:transfer and vf:transfer-all-rights actions.  Is there a relationship of the resource that is affected by vf:transfer-custody?

No.  We know that the accounting relationship (aka "ownership" in the current world) is useful directly on a resource.  We'd be very interested in finding out if you need more than that, such as the custodial relationship, which tends to be a shorter relationship; or other kinds of relationships that aren't in VF.  (In our older software, we supported a more complex model of user defined agent-resource relationships - which people used, but we haven't really seen it needed yet in projects using VF, and aren't sure it should be part of the model.  But that's why we are on version .5 :relaxed: .)

**Comment by fkleedorfer** [comment]
> Hmmm. I think c) is best, it is good for events to always have a quantity. And more consistent with moving part of a resource. I assume this scenario means you just want to change the currentLocation.

I agree, that's what I implemented.

**Comment by fkleedorfer** [comment]
> >  Is there a relationship of the resource that is affected by vf:transfer-custody?

> No. We know that the accounting relationship (aka "ownership" in the current world) is useful directly on a resource. We'd be very interested in finding out if you need more than that, such as the custodial relationship, which tends to be a shorter relationship; or other kinds of relationships that aren't in VF. (In our older software, we supported a more complex model of user defined agent-resource relationships - which people used, but we haven't really seen it needed yet in projects using VF, and aren't sure it should be part of the model. But that's why we are on version .5 ☺️ .)

When modeling transport it seems to be rather natural to differentiate between the owner and the custodian, and to have the information directly in the model, controlled by events. I am thinking about the prerequisites for actions so as to be able to reject actions that don't make sense. One such prerequisite could be that an agent cannot move/pickup/dropoff/transfer-custody/use/consume/modify(?)/accept(?)/raise(?)/lower(?) a resource unless the agent has custody. An agent cannot transfer/transfer-all-rights a resource unless they are primary accountable. That kind of thing. I think, a `vf:custodian` (resource->agent) relationship would provide that.

As a sidenote: I feel that this issue is connected to, or maybe obscured by, the topic of representing the same physical resource as different VF resources, controlled by different agents, let's say, each has their own data space - then, maybe, the custodian relationship may be  expressed at least in part by the fact that a resource enters the data space of a new agent.

**Comment by fkleedorfer** [comment]
### How are Resource Properties affected by events?
In the ontology and the docs, I see many properties that don't seem to be influenced by events, such as `vf:conformsTo`, `vf:containedIn`, `vf:lot`, `vf:stage`, `vf:state`, `vf:tracking identifier`, `vf:unitOfWork` (that last one is from the class diagram)

How do these properties find their way into the resource, given the principle that only events can affect resources, as @bhaugen said in #630.

I have some theories and some questions:
* **`vf:stage` and `vf:state`** are explained in the [Resources chapter](https://valueflo.ws/introduction/resources.html) and I believe I could model them that way. 
* **`vf:conformsTo`**: If I understand correctly, this relationship is a constant for a given resource - it will be there over its whole lifecycle (maybe a `vf:modify` can change that). So, does the `vf:produce` action set this property? If so, how? If not, how else does that happen?
* **`vf:lot`**: This is also explained in the resource chapter. 
    - How does the property get set on the resource? Is it a property of the event that is copied to the resource upon creation? 
    - Is this a property that can change over the lifetime of a resource? If so, how? 
    - It says that lot-based resources can be split up. How does that happen? Is this done with the `vf:move` action with a `vf:toResourceInventoriedAs` and the quantity to split off? If so, I'd have to adapt my understanding of `vf:move` because I thought the point is to change the `vf:currentLocation` of the resource. Alternatively, there is an action missing for taking part of a resource and making a new one of it; same for combining resources. 
    - Also, the vocabulary states that the range of the `vf:lot` relationship is the class [`dfc:ProductBatch`](http://www.virtual-assembly.org/DataFoodConsortium/BusinessOntology), but I guess the idea is that it's just some identifier, whatever you want, right? 
*  **`vf:containedIn`**: This seems to be a relationship that can change: put resource in other resource, take it out again. I don't see actions for that. I might need this for the transport, to put packages into trucks or into storage units, so I'd tentatively define `vf:putInto` and `vf:takeOutOf`, or whatever names you suggest. Unless, of course, there is a misunderstanding on my part.
* **`vf:tracking-identifier`**: 
    - Which action sets that?
    - Can it change?
    - Can it be removed?
    - Would it be too much to add two more actions, eg. `vf:setTrackingIdentifier` and `vf:removeTrackingIdentifier`? The alternative, I think, is to make it a property that can be set on any action and that overwrites whatever is set on the resource currently, but there would be no way to remove it.
* **`vf:unitOfWork`**:
    - What's that about?

In addtion to these questions: What is the semantics of the `vf:modify` action, i.e. how does it affect the resource?

### Regarding the shacl work
FYI, I've started the shacl program at https://github.com/fkleedorfer/valueflows-shacl. It comes with a number of unit tests and expected outputs for a) validating the example b) calculating inferences from the example (which is also done if the example is invalid) and c) validating the inferences. 

Cheers!

**Comment by fosterlynn** [comment]
>FYI, I've started the shacl program at https://github.com/fkleedorfer/valueflows-shacl. It comes with a number of unit tests and expected outputs

Exciting!  Very interested to see how this goes!

>vf:unitOfWork (that last one is from the class diagram)

Mistake in class diagram, it became unitOfEffort to account for both work and usage, will fix, thanks for catching.

>vf:conformsTo: If I understand correctly, this relationship is a constant for a given resource - it will be there over its whole lifecycle (maybe a vf:modify can change that). So, does the vf:produce action set this property? If so, how?

Yes, this is the type of resource, and is a constant over the lifecycle.  Yes it is set by a `produce` event, or perhaps `raise` (for beginning balances), or the receiving end of a `transfer`, anything that "creates" the resource in its scope or namespace.  `modify` should not change it, that is for modifications that do not change the identification of the resource or change its essence; if a process makes that many changes, I'd say better to use `consume` and `produce`.  And, it gets set by the `conformsTo` on the event that created the resource.

>What is the semantics of the vf:modify action, i.e. how does it affect the resource?

In addition to the above, think of `modify` for maintenance or repair.  But also for when a resource is created through defined stages (where `vf:stage` comes in from the process type).  We first came upon this for translations, which are created in a process sequence something like translate, copyedit, proofread, format.  It was cleaner and we thought more accurate to have one resource travel through those stages instead of having a separate resource (and resource specification) for each stage.  (Arguments could be made the other way, of course - do you have a translation if it hasn't been proofread?  That's why the logical identifier of the translation resource looks something like x-translation@translate, x-translation@copyedit, etc.)

>`vf:lot` `vf:trackingIdentifier` `vf:containedIn`

Yes, we've kind of cheated these in (as well as the resource note and others that don't really matter to the accounting side), somewhat discussed in #630 .  Implementations so far have just included the properties not explicitly set by an event in the event creation code by including resource properties in the parameters sent to the event creation code.  After that they are handled in the resource update code.  This doesn't seem right to people writing in newer functional or "monotonic" languages, such as yourself, understandably.  And while things like lot or tracking identifier don't matter to the accounting side, they do matter for some operational needs.  I'm leaning towards thinking we should clean that up in VF, to make it actually so that you can recreate all properties of a resource from re-running the economic events, which is not totally true now.  Or we could just push it farther in that direction, for the properties you are talking about, but not note, etc. This all might take a bit of thought, and a little discussion with a few people, so I'll kick that off.

>vf:lot: 

>    Is this a property that can change over the lifetime of a resource? If so, how?

>    It says that lot-based resources can be split up. How does that happen? Is this done with the vf:move action with a vf:toResourceInventoriedAs and the quantity to split off? If so, I'd have to adapt my understanding of vf:move because I thought the point is to change the vf:currentLocation of the resource. Alternatively, there is an action missing for taking part of a resource and making a new one of it; same for combining resources.

Lots (and tracking identifiers) were created because now people can't track and trace using value flows.  And it will probably be awhile, if ever.  And even then they will create useful short cuts, marked on the item itself, and be used in case any piece of a flow is missing for whatever reason.  Lots are particularly important in food and medicine, where they start with the "batch" being initially produced.  Lots are then most often split up into many resources, either as they are transferred a case or bottle or whatever at a time to different agents, or even as they are transformed to different resource specifications.  We understand a cow gets a lot identifier which follows all the cuts of meat, for example, so you can tell what cuts came from the same animal.

Yes, a `move` can move only part of a resource, say 50 of 100 cases, to a new location, so it is the way to split up a resource into 2 resources internal to the scoping agent, when a whole transportation process is not worth recording.  The lot would follow any divisions of the initial resource. So, note that if you aren't keeping track of locations, you can use `move` to do the split without changing `currentLocation` (since it wouldn't be in the data).  And yes, the quantity would be 50 in this example.

>Also, the vocabulary states that the range of the vf:lot relationship is the class dfc:ProductBatch, but I guess the idea is that it's just some identifier, whatever you want, right?

Good point.  We split that off into a separate entity to support food and anything else with expiration dates or other lot information worth recording.  But that isn't everything.  Maybe we should give an option for a ProductBatch or a string lot identifier?  

>vf:tracking-identifier:
>    Can it change?    Can it be removed?

Generally shouldn't be changed or removed, except for data entry mistake.  For sure if also physically identified on the resource, like a serial number.  Understand this doesn't necessarily help with validation logic, because of possibility of data entry mistake.

**Comment by fkleedorfer** [comment]
### Resource Creation

I am now trying to model `vf:produce` because I am interested in seeing how a resource can be created. If we stick with the idea that resources are only affected by events, this is the result I have so far:

Resources can only be created by `vf:produce`. However, `vf:produce` also adds to an existing resource if the `vf:resourceInventoriedAs` points to an existing one. (similar to `vf:move`, which may move one resource or move quantity from one resource to another, the latter if its `vf:toResourceInventoriedAs` relation is present).

This means that we have to differentiate 3 different cases for `vf:produce`:
1. there is no `[event] vf:resourceInventoriedAs [resource]` triple: This is the case if the event actually creates a resource. we can infer after this event that a new resource exists and link it to the event using `vf:resourceInventoriedAs`. The natural approach would be to set all resource values to the data provded in the event (ie.`vf:currentLocation`, `vf:onhandQuantity`, `vf:AccountingQuantity`, `vf:classifiedAs`, `vf:primaryAccountable`, at this point)
2. there is an `[event] vf:resourceInventoriedAs [resource]` triple: This is the case if
      a)  Case 1 happened and the inferences were added to the model, in which case the resulting resource is new (has no `vf:previousState`). For Validation, we have to compare the event properties with the resource properties as hinted to above.
      b)  The event merely adds quantity to an existing resource (it has a `vf:previousState`). In this case, for validation, we only have to compare some properties between the resource and the event (maybe location and classification), and we have to validate that the resources's next state properly reflects the change in quantity by the `vf:produce` event's. If the resource has no next state yet, the event is *current* and we have yet to infer the next resource state from the event and the resource, applying the increment rules instead of the creation rules from case 1.

If indeed `vf:produce` is the only way to create a resource, we'd also have to think about other data being set on the resource, like `vf:image`, or `skos:note` (which, by the way, I haven't seen an event-based way to change, yet, either). The other two possible interpretations of what I've learned about the model so far, with respect to resource creation, are:
* any event can create a resource. Just reference a not-yet-existing resource URI, and it's there
* the events that support `vf:toResourceInventoriedAs` can create a resource with that property pointing to a not-yet-existing resource.

Both of these interpretations incur complexity because they mean more exceptions, more valid combinations of events/resources, hence more and more complex code, more maintenance, and more bugs.

What if we defined (only for the SHACL program if needed), an action `vf:create`, that as the only action that may have no `vf:resourceInventoriedAs` (as in case 1 of ` vf:produce` above) and that all other actions MUST reference only preexisting resources? We could also define `vf:edit` for changing non-flow information, such as image, note, or other description info.

**Comment by fkleedorfer** [comment]
Ok, so in your response, I see that basically any event should be allowed to create a resource. So another way to remove the number of exceptions, one could differentiate between the create and add cases of actions by assigning different actions, e.g.  
* `vf:produce` -> `vf:produce` and `vf:produceNew`
* `vf:transfer` -> `vf:transfer`, `vf:transferTo`, and `vf:transferToNew` (also dealing with the optional "toResource here, which is quite a different case from the one without a "toResource", if you look at it)
* `vf:move` -> `vf:move`, `vf:moveTo`, `vf:moveToNew` (similar to the transfer)
and so on. 

Thoughts?

**Comment by fkleedorfer** [comment]
My hope that only `vf:produce` and `vf:consume` can change existence, btw is inspired by the action table in the [Flows Chapter](https://valueflo.ws/introduction/flows.html).

**Comment by fkleedorfer** [comment]
> > What is the semantics of the vf:modify action, i.e. how does it affect the resource?

> modify should not change it, that is for modifications that do not change the identification of the resource or change its essence;

> In addition to the above, think of modify for maintenance or repair. But also for when a resource is created through defined stages (where vf:stage comes in from the process type). We first came upon this for translations, which are created in a process sequence something like translate, copyedit, proofread, format. It was cleaner and we thought more accurate to have one resource travel through those stages instead of having a separate resource (and resource specification) for each stage. (Arguments could be made the other way, of course - do you have a translation if it hasn't been proofread? That's why the logical identifier of the translation resource looks something like x-translation@translate, x-translation@copyedit, etc.)

So `vf:modify` is for changing non-essential properties? I don't understand, though, why it has a incrementing effect, how it 'pairs' with `vf:accept` and why that one has a decrementing effect. 

As I am contemplating how to implement that any action can create resources (it looks very complicated), I am starting to like the idea of a `vf:create` - `vf:edit`- `vf:delete` lifecycle more and more, with `vf:create` being allowed to set any properties, but `vf:edit` only being allowed to touch the "non-essential" ones. All the other actions, on the other hand, operate on the essential ones. In this design, `vf:edit` could be used for most cases where `vf:modify` is used now.

**Comment by fosterlynn** [comment]
Some relatively quick replies.  And more thoughts later to your bigger questions and thoughts.

>So vf:modify is for changing non-essential properties? I don't understand, though, why it has a incrementing effect, how it 'pairs' with vf:accept and why that one has a decrementing effect.

The way it works is that a resource comes in to a process with `accept`, and out of that process with `modify`.  It doesn't have an increment / decrement for the accounting quantity, only the onhand quantity, which assumes it is not available while it is in the process.

>Resources can only be created by vf:produce.  [and some further discussion]

Just to make sure:  we are talking in this discussion about resource records being created, right?  That is sometimes confusing because you can sometimes see something is the same thing, but it is a "new" resource because it is an accounting artifact, with some reference to real world things, but not the same as that.  It is helpful to think about "stock" resources.

So I think the actions that can create a "stock" resource are: produce, all the transfers, move, and raise (when a beginning balance in a system).  

Note that of these, when they affect an existing resource, produce and raise will use the resourceInventoriedAs property; the rest will use the toResourceInventoriedAs property.  I'm not sure this is the most intuitive, but either way is potentially confusing; and only recently did we add toResourceInventoriedAs after long discussion about transfers, where produce and raise have been around much longer, and always just addressed resourceInventoriedAs.  So basically, toResourceInventoriedAs is only used when there is already a resourceInventoriedAs. 

Another thing to keep in mind is that sometimes inventory-able resources are not inventoried because they are not worth it, so there won't be an economic resource stored.  Often things that are easily seen for re-order, and/or a pain to measure, like a big pile of a certain size of nails.  In these cases, the resource specification is the most detailed information there will be in the event, and if a quantity is needed, it an only be gotten from events.  So, lack of a resourceInventoriedAs on an event does not always mean that a resource should be created.  (I think some of your suggestions speak to that.)

More later....

P.S. All of this discussion I will go back through and make some improvements in the doc!

**Comment by fkleedorfer** [comment]
### `vf:transfer` for resource with different accounting/onhand quantities

If we go with the principle that all events should (MUST?) have a quantity, what does that mean for a `vf:transfer` event (without a `vf:toResourceInventoriedAs`) in the case that its resource has different `vf:accountingQuantity` and `vf:onhandQuantity` values?

The meaning of `vf:transfer` is: give the thing to the other agent, nothing remains. So the effect should be that the resource gets `vf:primaryAccountable [receiver]` ( and `vf:custodian [receiver]`, I am toying with that), but nothing else about the resource should change.

One option would be to require `vf:onhandQuantity` and `vf:accountingQuantity` to be equal before such a transfer; that would be the clean thing to do - do you think that would cause problems down the line?

**Comment by fosterlynn** [comment]
>The meaning of vf:transfer is: give the thing to the other agent, nothing remains. 

In all cases, an event that decrements a resource can decrement part of it (assuming it is divisible).  But even if the resource goes to zero, it can still be the same identified "stock" resource, and more coming in might be expected that will raise the quantity of that same resource, even though the "touchable" items might be different ones in reality.  I think the "stock" nature of substitutable resources is one of the hardest things to understand in REA.

If you are transferring a famous painting say, and you keep track of all your paintings individually, then yes, nothing remains.  The new identifier of the painting in the scope of the new agent is most likely going to be different.  But I don't think we can say that absolutely either, in case there become more resources with kind of public non-agent-based identifiers on the web, or something like that.

Or for food where lot number is part of the logical identification, a resource would be gone when it is down to zero.

>If we go with the principle that all events should (MUST?) have a quantity, what does that mean for a vf:transfer event (without a vf:toResourceInventoriedAs) in the case that its resource has different vf:accountingQuantity and vf:onhandQuantity values?

Depending on the action, it can only in reality decrement what is onhand - like `consume`.  `transfer-all-rights` doesn't care what is onhand because only the rights to that quantity of resource are being transferred, thus the ability to change ownership "FOB destination", after the items have reached the owner-to-be.

Also, the identification of an inventoried resource is always optional, and does not really affect anything about how we think about the quantity.  Any agent might or might not specifically identify inventoried resources.  They just as easily could identify "bike pedals part XYZ" just with a resource specification, and not worry even if they have their bike pedals part XYZ in two bins in different rooms.  Just depends on what they need operationally.  (There is a section on identifying resources on that page in the doc.) 

Also, on a transfer, since VF is based on the "collaboration space" between agents, an agent could internally have an identifier for a resource, but it doesn't show on the transfer event because it is not known by the other agent.  (I'm not sure if we have actually resolved this in Holochain, where a record is identified technically by its hash, so you can't add a property to it later without changing the identifier.)

On a transfer event, each agent will be responsible for decrementing or incrementing their side of the event, unless they are closely related enough to be using the same instance of the computer system.

P.S. In terms of validation rules, we started defining them here https://github.com/valueflows/vf-apps/issues.  They are incomplete, and we haven't gotten them out of issues into a more proper document yet.

P.P.S. @bhaugen as the person with most experience on the ground in manufacturing and supply chains, as always feel free to correct me or clarify!

**Comment by fkleedorfer** [comment]
Thanks! To clarify, I meant specifically if you have a `vf:transfer` (not `vf:transfer-all-rights` or `vf:transfer-custody`) that references exactly one resource (no `vf:toResourceInventoriedAs`).

My understanding would be that that changes nothing about the resource but the `vf:primaryAccountable`. However, if the event has a quantity, the question is, how that needs to be factored in. You said, you cannot transfer more than is onhand, on the other hand I'd add you cannot transfer more than you are accounting for. 

Let's say you have 5 lawnmowers (onhand=5), but one of them isn't yours(accounting=4), you are just holding on to it for someone (say, because you sold it (using transfer-all-rights) and the person needs to come pick it up).  That could be a situation in which you'd want to forbid a `vf:transfer` as described above.

**Comment by fosterlynn** [comment]
>My understanding would be that that changes nothing about the resource but the vf:primaryAccountable.

In this case, I don't think we can assume that, because of the optionality of any inventoried resource being recorded.  The most likely is that it is going out of the namespace.  Even if not, it could work that way, still having the decrement and the creation or a new resource or whatever. 

>You said, you cannot transfer more than is onhand, on the other hand I'd add you cannot transfer more than you are accounting for.

Yes, makes perfect sense.

**Comment by pospi** [comment]
@fkleedorfer just thought I'd reference this here for you as it's the canonical reference in terms of code AFAIK. Where you see references to the `e` input arguments vs `self` are where values are being overwritten vs retained. https://github.com/valueflows/valueflows/issues/637#issuecomment-636455874

**Comment by bhaugen** [comment]
I'm gonna try to explain how to think about the differences and relationships between vf:EconomicResource and physical stuff. If this gets good enough, maybe it should be posted in the documentation somewhere.

In general, think of an EconomicResource as an accounting record, not an exact image of the physical stuff. So do what is useful for the agents involved for their accounting purposes. Which will be very situational. And usually does not need to be as detailed as all the possible properties of the physical stuff.

The reasons that you might want to account for economic resources include:
* How much of this Resource Specification do we have on hand for some offer or need?
    * And where is it? (which is situational, see next paragraph)
* What is the value of our inventory for tax or economic analysis purposes? (Which might want different valuation algorithms and thus different values...)

If an organization has a controlled inventory with an inventory management staff, then "where is it?" might not be any of a VF app's business, because the inventory management staff moves things around at will and knows where things are and does not record the movements because it's not worth the effort. Or the inventory might be in an automated warehouse and only the warehouse robot knows where it is and the information is usually in a separate database.

One actual countable physical resource, like a bicycle, might have a serial number (a unique identifier which would go into the vf:trackingIdentifier property). In which case, almost always, the physical resource will have a visible serial number marked on it.

In some cases, records of transfers or transportation of serialized resources are accompanied by a list of the serial numbers involved.

That bicycle might be stored in a bin with many other bicycles which might be a useful vf:EconomicResource for accounting purposes, including how many bicycles do we have on hand? 

One batch of measurable but not usually countable physical resources like liquids or powders (e.g. beer or flour) might have a lot or similar batch identifier that go into vf:lot and/or vf:trackingIdentifier. (You'd do one or the other or both depending on requirements: the lot property requires a separate dfc:ProductBatch record).

A lot identifier is also usually marked visibly on each physical resource container (like packages) of resources that were produced in the same batch. 

So where do you stop in the possible levels of detail? Do you need to get into, for example, molecular properties? 

If you are [Impact Bioenergy](http://impactbioenergy.com/) and are producing fertilizers, you might want to do some measurements like these: https://uwosh.edu/eric/compost-testing-packages/

If you are producing beer, maybe all you need is the alcohol percentage. For food, maybe the nutritional measurements that are printed on a package label. And both of those measurements are standardized for a resource specification, not measured on every batch.

For soap, maybe none of the above.

Because the observations of Economic Resources are recorded from the viewpoint of one Economic Agent, when some resources change from the ownership or custody to the ownership or custody of another Economic Agent, they become different Economic Resources, although they are the same physical stuff.

So, to summarize, record vf:EconomicResources in the level of detail that is useful for knowing what an agent has on hand and optionally where it is, and what is the value of an agent's inventory, and what are the properties of the resources that might be useful for particular offers or needs.

**Comment by orthecreedence** [comment]
> No. We know that the accounting relationship (aka "ownership" in the current world) is useful directly on a resource. We'd be very interested in finding out if you need more than that, such as the custodial relationship, which tends to be a shorter relationship; or other kinds of relationships that aren't in VF.

Just want to drop by and say that I'm implementing an `in_custody_of` field on the type containing `EconomicResource` because I'd like the system to track custody. I think it would be a natural fit to have this built into the VF resource data model (and I almost expected it and was looking for it once I started implementing the resource portion of events). So I'd vote :+1: on a custody-tracking field.

**Comment by fkleedorfer** [comment]
@orthecreedence So you did not like `custodian` - I get that. When I looked that up (I am not a native speaker), my association was a monk taking good care of some ancient treasure ;-) I was also thinking of `inCustodyOf`, but I wanted to go with the same idea as `primaryAccountable` - the name of the role. Given that the latter actually wasn't so intuitive for me, `inCustodyOf` may be the better choice, but it's a bit inconsistent. We could change `primaryAccountable` to `ownedBy`, of course.

I am now experimenting with planning algorithms and I am not sure if all the adaptations that seem warranted in our context have a chance to make it back into VF. I guess the difference stems from the global perspective of WoN as opposed to the organizational perspective that VF seems to have guided VF development so far.

Anyway, with our current design, the custodial relationship is absolutely necessary for planning - e.g. you cannot transport something if you don't actually have a truck. If you own one, but you lent it to someone, the planner has to be able to account for that. You can borrow it from someone, else but only if that person actually has it and owns it.

**Comment by orthecreedence** [comment]
@fkleedorfer To be fair, I implemented `in_custody_of` before I saw this issue =]. If we are keeping the naming consistent, then I would agree "custodian" might be a better option. Either way, I support the existence of the field.

**Comment by bhaugen** [comment]
@fkleedorfer 
> I am now experimenting with planning algorithms and I am not sure if all the adaptations that seem warranted in our context have a chance to make it back into VF. I guess the difference stems from the global perspective of WoN as opposed to the organizational perspective that VF seems to have guided VF development so far.

I see a bunch of separate topics enmeshed in your very interesting paragraph. I'll start with planning and planning algorithms because I've written about that topic in the VF context recently, and am curious if the planning algorithms you have in mind are anything like those I've written about. Here are some links:
* https://github.com/valueflows/vf-apps-planning pretty sparse so far, but
* https://github.com/valueflows/vf-apps-planning/issues/1 started.
* https://github.com/valueflows/valueflows/issues/284 Budgets, estimates, forecasts, and plans
* https://github.com/valueflows/valueflows/issues/472 Planning scenarios

I'd like to follow up with these other topics, maybe in separate issues?

>  if all the adaptations that seem warranted in our context have a chance to make it back into VF

Or, more generally, how should adaptations from different contexts make it back into VF? VF emerged from a long series of participatory action research projects and now is going back into new projects which are deliberately using VF. We (me and @fosterlynn , anyway) see this as a spiral, learning from particular experiments, summarizing into more-general conclusions (making it back into VF), and then going back into practice (more particular experiments).

>  the difference stems from the global perspective of WoN as opposed to the organizational perspective that VF seems to have guided VF development so far.

@fkleedorfer is it ok if I start the "adaptations making it back into VF" issue, and would you like to start a "global vs organizational perspective" issue?

**Comment by fkleedorfer** [comment]
> One could split this out into different economic resources for the owner, with a different custodian on each one, but that gets very messy, and imo not worth it.

@fosterlynn I think messy or not it is a question of how you enable this in the UI, but in terms of data modelling it seems pretty clean to me, actually. As long as we are talking about physical resources that cannot possilbly be in the custody of more than one actor, this is the approach I'd follow. Same with intangibles that can be restricted to one actor (e.g. a per-user license).

**Comment by almereyda** [comment]
We have moved the ValueFlows organization from GitHub to https://lab.allmende.io/valueflows.

This issue has been closed here, and all further discussion on this issue can be done at

> https://lab.allmende.io/valueflows/valueflows/-/issues/636.

If you have not done so, you are very welcome to register at https://lab.allmende.io and join the ValueFlows organization there.
